File: src/main/java/com/kamikazejam/datastore/base/log/LoggerService.java
package com.kamikazejam.datastore.base.log;

import com.kamikazejam.datastore.util.Color;
import org.bukkit.Bukkit;
import org.bukkit.plugin.Plugin;
import org.jetbrains.annotations.NotNull;

import java.util.logging.Level;

@SuppressWarnings("unused")
public abstract class LoggerService {

    // Abstraction
    public abstract String getLoggerName();

    public abstract Plugin getPlugin();

    public abstract boolean isDebug();

    // Public Methods
    public void info(@NotNull String msg) {
        logToConsole(msg, Level.INFO);
    }

    public void info(@NotNull Throwable throwable) {
        logToConsole(throwable.getMessage(), Level.INFO);
        throwable.printStackTrace();
    }

    public void info(@NotNull Throwable throwable, @NotNull String msg) {
        logToConsole(msg + " - " + throwable.getMessage(), Level.INFO);
        throwable.printStackTrace();
    }

    public void debug(@NotNull String msg) {
        if (!isDebug()) {
            return;
        }
        logToConsole(msg, Level.FINE);
    }

    public void warn(@NotNull String msg) {
        logToConsole(msg, Level.WARNING);
    }

    public void warn(@NotNull Throwable throwable) {
        logToConsole(throwable.getMessage(), Level.WARNING);
        throwable.printStackTrace();
    }

    public void warn(@NotNull Throwable throwable, @NotNull String msg) {
        logToConsole(msg + " - " + throwable.getMessage(), Level.WARNING);
        throwable.printStackTrace();
    }

    public void warning(@NotNull String msg) {
        this.warn(msg);
    }

    public void warning(@NotNull Throwable throwable) {
        this.warn(throwable);
    }

    public void warning(@NotNull Throwable throwable, @NotNull String msg) {
        this.warn(throwable, msg);
    }

    public void severe(@NotNull String msg) {
        logToConsole(msg, Level.SEVERE);
    }

    public void severe(@NotNull Throwable throwable) {
        logToConsole(throwable.getMessage(), Level.SEVERE);
        throwable.printStackTrace();
    }

    public void severe(@NotNull Throwable throwable, @NotNull String msg) {
        logToConsole(msg + " - " + throwable.getMessage(), Level.SEVERE);
        throwable.printStackTrace();
    }

    public void error(@NotNull String msg) {
        this.severe(msg);
    }

    public void error(@NotNull Throwable throwable) {
        this.severe(throwable);
    }

    public void error(@NotNull Throwable throwable, @NotNull String msg) {
        this.severe(throwable, msg);
    }

    public void logToConsole(String msg, Level level) {
        String plPrefix = "[" + this.getPlugin().getName() + "] ";
        if (level == Level.FINE) {
            Bukkit.getConsoleSender().sendMessage(Color.t("&7[DEBUG] " + plPrefix + msg));
        } else if (level == Level.INFO) {
            Bukkit.getConsoleSender().sendMessage(Color.t(plPrefix + msg));
        } else if (level == Level.WARNING) {
            Bukkit.getConsoleSender().sendMessage(Color.t("&e[WARNING] " + plPrefix + msg));
        } else if (level == Level.SEVERE) {
            Bukkit.getConsoleSender().sendMessage(Color.t("&c[SEVERE] " + plPrefix + msg));
        }
    }
}

File: src/main/kotlin/com/kamikazejam/datastore/base/log/LoggerService.kt
package com.kamikazejam.datastore.base.log

import com.kamikazejam.datastore.util.Color
import org.bukkit.Bukkit
import org.bukkit.plugin.Plugin
import java.util.logging.Level

@Suppress("unused", "MemberVisibilityCanBePrivate")
abstract class LoggerService {
    // Abstraction
    abstract val loggerName: String

    abstract val plugin: Plugin

    abstract val isDebug: Boolean

    // Public Methods
    fun info(msg: String) {
        logToConsole(msg, Level.INFO)
    }

    fun info(throwable: Throwable) {
        logToConsole(throwable.message, Level.INFO)
        throwable.printStackTrace()
    }

    fun info(throwable: Throwable, msg: String) {
        logToConsole(msg + " - " + throwable.message, Level.INFO)
        throwable.printStackTrace()
    }

    fun debug(msg: String) {
        if (!isDebug) {
            return
        }
        logToConsole(msg, Level.FINE)
    }

    fun warn(msg: String) {
        logToConsole(msg, Level.WARNING)
    }

    fun warn(throwable: Throwable) {
        logToConsole(throwable.message, Level.WARNING)
        throwable.printStackTrace()
    }

    fun warn(throwable: Throwable, msg: String) {
        logToConsole(msg + " - " + throwable.message, Level.WARNING)
        throwable.printStackTrace()
    }

    fun warning(msg: String) {
        this.warn(msg)
    }

    fun warning(throwable: Throwable) {
        this.warn(throwable)
    }

    fun warning(throwable: Throwable, msg: String) {
        this.warn(throwable, msg)
    }

    fun severe(msg: String) {
        logToConsole(msg, Level.SEVERE)
    }

    fun severe(throwable: Throwable) {
        logToConsole(throwable.message, Level.SEVERE)
        throwable.printStackTrace()
    }

    fun severe(throwable: Throwable, msg: String) {
        logToConsole(msg + " - " + throwable.message, Level.SEVERE)
        throwable.printStackTrace()
    }

    fun error(msg: String) {
        this.severe(msg)
    }

    fun error(throwable: Throwable) {
        this.severe(throwable)
    }

    fun error(throwable: Throwable, msg: String) {
        this.severe(throwable, msg)
    }

    fun logToConsole(msg: String?, level: Level) {
        val plPrefix = "[" + plugin.name + "] "
        if (level === Level.FINE) {
            Bukkit.getConsoleSender().sendMessage(Color.t("&7[DEBUG] $plPrefix$msg"))
        } else if (level === Level.INFO) {
            Bukkit.getConsoleSender().sendMessage(Color.t(plPrefix + msg))
        } else if (level === Level.WARNING) {
            Bukkit.getConsoleSender().sendMessage(Color.t("&e[WARNING] $plPrefix$msg"))
        } else if (level === Level.SEVERE) {
            Bukkit.getConsoleSender().sendMessage(Color.t("&c[SEVERE] $plPrefix$msg"))
        }
    }
}
